### PR TITLE
Add zonal stats reporting to visualizations

### LINKED ISSUE #21: Add zonal statistics
Being able to calculate stats based on a polygon from either selected polygon from a user shapefile or from a user-defined region (using selection capabilities).

### PR DESCRIPTION
**Pull request by TaylorMutch**
This PR adds the ability to perform zonal statistics on a terrain visualization, allowing a user to draw a box or a customized polygon feature onto the terrain. This PR also enhances the ESRIGridAscii data plugin to use `rasterio` for improved I/O performance, and it enhances the NetCDFDataPlugin to use `clover` for consistent I/O interactions. Resolves #21.

Remaining TODOs:
- [x] Investigate improved box selection for large terrains (i.e. via BoundingBox plane clipping trick)
- [ ] Cleanup, documentation.

### CODE REVIEW

File: plugins/esri_grid_ascii/main.py [hunk 1]
```diff
@@ -115,93 +111,65 @@ def load_data(self):
...
-            else:
-                grid = self.get_data("")
-                stats.min_value = float(grid.min())
-                stats.max_value = float(grid.max())
-
-            stats.nodata_value = self._header['nodata_value']
+            maxs = []
+            mins = []
+            steps = self.time_info.timestamps if self.time_info.is_temporal else [0]
+            for step in steps:
```
Comment (nikmolnar) [inline plugins/esri_grid_ascii/main.py]: Untested, but this may work:

```python
stats.min_value = min(float(self.get_data('', step).nanmin()) for step in steps)
# same for max
```

File: plugins/netcdf/main.py [hunk 2]
```diff
@@ -92,42 +100,31 @@ def is_valid_file(path):
...
-        return self._name
-
     def get_data(self, variable, date=None):
-        with Dataset(self.path, 'r') as ds:
-            data = ds.variables[variable][:][self._temporal_info.timestamps.index(date)]
-        return data
+        if date is None:
+            date = Timeline.app().current
+
+        if not (variable == self._current_variable):
```
Comment (nikmolnar) [inline plugins/netcdf/main.py]: `if variable != self._current_variable`

File: plugins/terrain_and_color/main.py [hunk 3]
```diff
@@ -588,3 +591,83 @@ def get_identify_detail(self, point):
...
+                            zones.append(feat)
+
+                    # Retrieve zonal stats for this raster
+                    result = zonal_stats(zones, raster, affine=affine, nodata=nodata, add_stats=self.zonal_stats)
+                    for j, row in enumerate(result):
+                        row['Name'] = "{} (Zone {})".format(plugin.data_name, zones[j].get('id'))
+                    results.append(result)
+        return results
+
+    def get_zonal_stats_from_feature(self, feature: Dict) -> List[Optional[Dict]]:
```
Comment (nikmolnar) [inline plugins/terrain_and_color/main.py]: GeoJSON doesn't seem like the best in-memory representation to pass around geometries. Wouldn't shapely objects or such be more appropriate?

File: plugins/terrain_and_color/main.py [hunk 4]
```diff
@@ -588,3 +591,83 @@ def get_identify_detail(self, point):
         self.refresh()
 
         return None
+
+    def get_zonal_stats_from_point(self, point: Vector3) -> List[Optional[Dict]]:
+        results = []
+        if self.boundary_data:
+            for i, data_role in enumerate(self.data_roles):
+                dtype, _ = data_role
+                if dtype is DataPlugin.RASTER:
```
Comment (nikmolnar) [inline plugins/terrain_and_color/main.py]: Since this is plugin-specific and we know the data roles, couldn't we just iterate over those roles specifically and never have to test this?

Comment (nikmolnar) [reply plugins/terrain_and_color/main.py]: For that matter couldn't you also iterate over the data instead of roles? For example:

```python
for plugin in (x for x in (self.terrain_data, self.attribute_data) if x is not None):
    ...
```

File: source/vistas/core/graphics/overlay.py [hunk 5]
```diff
@@ -15,7 +15,7 @@ class Overlay:
     shader = None
 
     def __init__(self, canvas):
-        self._buttons = []
```
Comment (nikmolnar) [inline source/vistas/core/graphics/overlay.py]: I was using `_buttons` here for a very specific reason; namely that `_buttons` should never be modified directly, because of other stuff that needs to happen when buttons are added or removed.

File: source/vistas/ui/controls/gl_canvas.py [hunk 6]
```diff
@@ -73,19 +84,57 @@ def OnMotion(self, event: wx.MouseEvent):
...
+                wx.PostEvent(self.GetParent(), select_event)
+                self.selection_mode = None
+                self.selection_controls.hide_optional_buttons()
+            self.Refresh()
+        event.Skip()
+
+    def OnLeftDown(self, event: wx.MouseEvent):
+        if self.selection_mode and self.selection_mode is GLSelectionControls.POLY:
+
+            # Edge case where the user clicked on the overlay while in POLY mode
```
Comment (nikmolnar) [inline source/vistas/ui/controls/gl_canvas.py]: You might be able to just test for `self.overlay.target` E.g.

```python
if self.overlay.target:
    event.Skip()
    return
```

File: source/vistas/core/graphics/overlay.py (hunk 5 above)
Comment (TaylorMutch) [reply source/vistas/core/graphics/overlay.py]: Ah, I remember why I made this change - it was because I thought I would have to make a change inside of `GLSelectionControls`, but it ended up not happening. I'll revert it.

File: plugins/esri_grid_ascii/main.py (hunk 1 above)
Comment (TaylorMutch) [reply plugins/esri_grid_ascii/main.py]: Almost works, reason for the loop is because I need to filter out the `nodata_value` in the calculation, and it just makes a bit more sense to do it in the loop. Fun one-liner though if no masking is needed.

File: plugins/terrain_and_color/main.py (hunk 4 above)
Comment (TaylorMutch) [reply plugins/terrain_and_color/main.py]: Ah, yes we do that.

If we were having this logic outside of the plugin though, this would make sense. Make it so it doesn't have to be re-implemented for each plugin, but since this is the only plugin that does the work, then maybe it's okay for now.

File: plugins/terrain_and_color/main.py (hunk 3 above)
Comment (TaylorMutch) [reply plugins/terrain_and_color/main.py]: Could be. I'll test and see if it is okay. Main reason for the GeoJSON interface is so that I could swap in the coordinates for the same features, but `rasterstats` might take the `shapely` objects just fine.

File: plugins/esri_grid_ascii/main.py (hunk 1 above)
Comment (nikmolnar) [reply plugins/esri_grid_ascii/main.py]: That's what `nanmim` _should_ take care of...

Comment (TaylorMutch) [reply plugins/esri_grid_ascii/main.py]: It should if the input has `numpy.nan`s in there, but the `rasterio` read isn't masked (currently):
```
    with rasterio.open(path) as src:
        self._current_grid = src.read(1)
```
It just returns the data in the clear without a mask. We could add a 'mask=True` parameter for Raster-based plugins to do this, since for plugins where we want to create a terrain out of the data, we don't want it masked since we handle that transformation ourselves.

File: plugins/terrain_and_color/main.py (hunk 3 above)
Comment (TaylorMutch) [reply plugins/terrain_and_color/main.py]: Works fine.

File: plugins/esri_grid_ascii/main.py (hunk 1 above)
Comment (nikmolnar) [reply plugins/esri_grid_ascii/main.py]: It seems like where appropriate, data should be masked. Using numpy, any code using the data that doesn't want to the mask for some reason or another can ignore it.

**Comment by nikmolnar** [comment]
The selection moves with respect to the terrain. E.g., this is the same selection rectangle:

screen shot 2017-09-23 at 12 03 38 pm
screen shot 2017-09-23 at 12 03 44 pm

Perhaps we'd be better served by rendering the selection _on_ the terrain itself?

**Comment by nikmolnar** [comment]
I can't seem to do more than one zonal statistics operation. When I click the rectangle select to do another one, I just get:

screen shot 2017-09-23 at 12 05 34 pm

Clicking and dragging on the terrain just rotates the view.

**Comment by nikmolnar** [comment]
We should round zonal stats (for that matter identify and maybe a few other places) to most significant digit for display.

https://stackoverflow.com/questions/3410976/how-to-round-a-number-to-significant-figures-in-python

**Comment by nikmolnar** [comment]
Unrelated to this PR: we really need to add tooltip support to the overlay buttons.

**Comment by TaylorMutch** [comment]
@nikmolnar I added the changes we discussed. Feel free to give this another look as time permits.

**Comment by nikmolnar** [comment]
This is now giving me the "At least 3  pointers are required to do zonal statistics!" error as soon as I click on the rectangle selection tool.
